bindings/python/CONTRIBUTING.md

```diff
@@ -0,0 +1,790 @@
+cdef vector[mata_nfa.CMove].iterator it = transitions_list.begin()
+cdef vector[mata_nfa.CMove].iterator end = transitions_list.end()
+# ^-- define begin and end iterators
+transsymbols = []
+while it != end:
+#     ^-- comparsion of two iterators
+    t = Move(
+        dereference(it).symbol,
+        dereference(it).targets.ToVector()
+        # ^-- to access the value of iterator you need to dereference
+    )
+    postinc(it)
+    # ^-- ++it
+    transsymbols.append(t)
+return transsymbols
+```
+
+## Defining type definition
+
+* **Problem**: You want to create some custom type or type alias.
+* **Solution**: You simply define `ctypedef` in `pxd` file. This can then be cimported from module.
+
+```cython
+ctypedef umap[EpsilonDepth, TransitionSequence] EpsilonDepthTransitions
+```
+
+## Calling functions of shared pointer object
+
+* **Problem**: You need to work with shared pointers.
+* **Solution**: You cimport `shared_ptr` type, which can be used as any other type. For dereferencing
+  or working with the object use `.get()` function which returns the pointer.
+
+```cython
+from libcpp.memory cimport shared_ptr
+lhs.thisptr.get().add_state()
+```
+
+## Supporting Python for loops
+
+* **Problem**: You want iterate over your structure in Python `for` loops.
+* **Solution**: The standard function should support this by default (i.e., `std::` containers),
+  your custom classes needs to support basic iterators (`begin()`, `end()`, etc.).
+
+
+```cython
+cdef vector[CTrans] c_transitions = self.thisptr.get().get_transitions_to(state_to)
+trans = []
+for c_transition in c_transitions:
+#   ^-- classic for each loop over std::vector<Trans> does everything behind the scenes
+    trans.append(Trans(c_transition.src, c_transition.symb, c_transition.tgt))
+```
+
+## Working with strings
+
+* **Problem**: You want to work with strings and pass them between binding and wrapped library.
+* **Solution**: You need to use `encode()` when passing strings to C/C++ functions, and `decode()`,
+  when recieving results.
+
+```cython
+k.encode('utf-8')
+# ^- when passing strings to C/C++ library you need to encode
+result.decode('utf-8')
+# ^- when retrieving results from C/C++ you need to decode
+```
+
+## Helping Cython to infer types
+
+* **Problem**: Cython fails to infer types for the variables and cannot compile the binding.
+* **Solution**: You need to help Cython a little: add types to parameters, add types to cdefined 
+  variables, or manually retype the variable.
+
+```cython
+(<mata_nfa.Nfa>noodle_segment).thisptr = c_noodle_segment
+# ^-- tell Cython that this variable is of given type
+objects: List[subprocess.Popen] = []
+# ^-- tell Cython the type of the Python object
+cdef mata_nfa.stringstream* output_stream
+# ^-- define variable as C/C++ type
+def union(cls, Nfa lhs, Nfa rhs):
+    pass
+#              ^-- tell Cython types of parameters
+```
+
+## Using pre- and post-increments (`var++` and `++var`)
```

Comment:
Move this section above section `Working with iterators` where `postincrement` is already used.